File: src/agent/orchestrator/sandbox-runner.ts
/**
 * Sandbox Runner
 *
 * Provides sandboxed execution for orchestrator operations.
 * When sandbox is enabled, commands run inside a container with the workspace mounted.
 * Falls back to host execution when sandbox is disabled or unavailable.
 */
import * as BunContext from "@effect/platform-bun/BunContext";
import { Effect } from "effect";
import { spawn } from "node:child_process";
import type { SandboxConfig } from "../../tasks/schema.js";
import {
  runInContainer,
  isContainerAvailable,
  autoDetectLayer,
  type ContainerConfig,
  type ContainerRunOptions,
} from "../../sandbox/index.js";
import {
  createSandboxHudAdapter,
  type SandboxHudAdapter,
} from "../../sandbox/hud-adapter.js";
import {
  createCredentialMount,
  cleanupCredentialMount,
  type CredentialMount,
} from "../../sandbox/credentials.js";
import type {
  HudMessage,
  ExecutionContext,
} from "../../hud/protocol.js";
import * as nodeFs from "node:fs";

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

export interface SandboxRunnerConfig {
  /** Sandbox configuration from project.json */
  sandboxConfig: SandboxConfig;
  /** Working directory (repo root) to mount as /workspace */
  cwd: string;
  /** Emit events for logging/debugging */
  emit?: ((event: SandboxRunnerEvent) => void) | undefined;
  /** Emit HUD messages for streaming output to UI */
  emitHud?: ((message: HudMessage) => void) | undefined;
  /** Execution context for UI grouping (default: "verification") */
  context?: ExecutionContext | undefined;
}

export type SandboxRunnerEvent =
  | { type: "sandbox_check_start" }
  | { type: "sandbox_available"; backend: string }
  | { type: "sandbox_unavailable"; reason: string }
  | { type: "sandbox_command_start"; command: string[]; inContainer: boolean }
  | { type: "sandbox_command_complete"; command: string[]; exitCode: number; durationMs: number }
  | { type: "sandbox_fallback"; reason: string };

export interface CommandResult {
  exitCode: number;
  stdout: string;
  stderr: string;
  /** Whether the command ran in a container or on host */
  sandboxed: boolean;
}

interface SandboxBackendRunOptions {
  command: string[];
  cwd: string;
  sandboxConfig: SandboxConfig;
  env?: Record<string, string>;
  hudAdapter?: SandboxHudAdapter;
}

interface SandboxBackend {
  name: string;
  sandboxed: boolean;
  run: (
    options: SandboxBackendRunOptions,
  ) => Effect.Effect<CommandResult, Error, never>;
}

// ─────────────────────────────────────────────────────────────────────────────
// Default Image
// ─────────────────────────────────────────────────────────────────────────────

const DEFAULT_SANDBOX_IMAGE = "oven/bun:latest";

// ─────────────────────────────────────────────────────────────────────────────
// Sandbox Availability Check
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Check if sandbox execution is available.
 * Returns true if:
 * 1. sandbox.enabled is true in config
 * 2. sandbox.backend is not "none"
 * 3. A container backend is actually available on the system
 */
export const checkSandboxAvailable = (
  config: SandboxConfig,
  emit?: SandboxRunnerConfig["emit"]
): Effect.Effect<boolean, never, never> =>
  Effect.gen(function* () {
    emit?.({ type: "sandbox_check_start" });

    // Check if explicitly disabled
    if (config.enabled === false) {
      emit?.({ type: "sandbox_unavailable", reason: "sandbox.enabled is false" });
      return false;
    }

    // Check if backend is explicitly "none"
    if (config.backend === "none") {
      emit?.({ type: "sandbox_unavailable", reason: "sandbox.backend is 'none'" });
      return false;
    }

    // Check if container backend is available
    const available = yield* Effect.provide(
      isContainerAvailable(),
      autoDetectLayer
    ).pipe(
      Effect.catchAll(() => Effect.succeed(false))
    );

    if (available) {
      emit?.({ type: "sandbox_available", backend: config.backend ?? "auto" });
    } else {
      emit?.({ type: "sandbox_unavailable", reason: "no container backend available" });
    }

    return available;
  });

// ─────────────────────────────────────────────────────────────────────────────
// Container Config Builder
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Build ContainerConfig from SandboxConfig and working directory.
 */
export const buildContainerConfig = (
  sandboxConfig: SandboxConfig,
  cwd: string,
  options?: {
    env?: Record<string, string>;
    volumeMounts?: string[];
  },
): ContainerConfig => ({
  image: sandboxConfig.image ?? DEFAULT_SANDBOX_IMAGE,
  workspaceDir: cwd,
  workdir: "/workspace",
  ...(sandboxConfig.memoryLimit ? { memoryLimit: sandboxConfig.memoryLimit } : {}),
  ...(sandboxConfig.cpuLimit ? { cpuLimit: sandboxConfig.cpuLimit } : {}),
  ...(sandboxConfig.timeoutMs ? { timeoutMs: sandboxConfig.timeoutMs } : {}),
  ...(options?.env ? { env: options.env } : {}),
  ...(options?.volumeMounts ? { volumeMounts: options.volumeMounts } : {}),
  autoRemove: true,
});

// ─────────────────────────────────────────────────────────────────────────────
// Host Execution
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Run a command on the host (no sandbox) with optional streaming callbacks.
 */
const runOnHostWithCallbacks = (
  command: string[],
  cwd: string,
  options?: {
    timeoutMs?: number | undefined;
    onStdout?: (chunk: string) => void;
    onStderr?: (chunk: string) => void;
    env?: Record<string, string>;
  }
): Effect.Effect<CommandResult, Error, never> =>
  Effect.async((resume) => {
    // Join command array into shell command string
    const cmdString = command.join(" ");
    const proc = spawn(cmdString, {
      cwd,
      shell: true,
      stdio: ["pipe", "pipe", "pipe"],
      env: options?.env
        ? { ...process.env, ...options.env }
        : process.env,
    });

    let stdout = "";
    let stderr = "";
    let finished = false;

    // Set timeout
    const timeoutMs = options?.timeoutMs ?? 120000;
    const timeoutHandle = setTimeout(() => {
      if (!finished) {
        proc.kill("SIGKILL");
        finished = true;
        resume(Effect.fail(new Error(`Host command timed out after ${timeoutMs}ms`)));
      }
    }, timeoutMs);

    proc.stdout?.on("data", (data: Buffer) => {
      const chunk = data.toString("utf-8");
      stdout += chunk;
      options?.onStdout?.(chunk);
    });

    proc.stderr?.on("data", (data: Buffer) => {
      const chunk = data.toString("utf-8");
      stderr += chunk;
      options?.onStderr?.(chunk);
    });

    proc.on("error", (err: Error) => {
      if (!finished) {
        clearTimeout(timeoutHandle);
        finished = true;
        resume(Effect.fail(new Error(`Host command failed: ${err.message}`)));
      }
    });

    proc.on("close", (code: number | null) => {
      if (!finished) {
        clearTimeout(timeoutHandle);
        finished = true;
        resume(Effect.succeed({
          exitCode: code ?? 1,
          stdout,
          stderr,
          sandboxed: false,
        }));
      }
    });

    // Cleanup on abort
    return Effect.sync(() => {
      if (!finished) {
        clearTimeout(timeoutHandle);
        proc.kill("SIGKILL");
      }
    });
  });

/**
 * Run a command on the host (no sandbox).
 * @deprecated Use runOnHostWithCallbacks for streaming support
 */
const runOnHost = (
  command: string[],
  cwd: string,
  timeoutMs?: number
): Effect.Effect<CommandResult, Error, never> =>
  runOnHostWithCallbacks(command, cwd, { timeoutMs });

// ─────────────────────────────────────────────────────────────────────────────
// Sandboxed Execution
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Run a command in a sandbox container.
 */
const runInSandboxContainer = (
  command: string[],
  containerConfig: ContainerConfig,
  streamingOptions?: ContainerRunOptions
): Effect.Effect<CommandResult, Error, never> =>
  Effect.gen(function* () {
    const result = yield* Effect.provide(
      runInContainer(command, containerConfig, streamingOptions),
      autoDetectLayer
    ).pipe(
      Effect.mapError((e) => new Error(`Container execution failed: ${e.message}`))
    );

    return {
      exitCode: result.exitCode,
      stdout: result.stdout,
      stderr: result.stderr,
      sandboxed: true,
    };
  });

// ─────────────────────────────────────────────────────────────────────────────
// Backend Implementations
// ─────────────────────────────────────────────────────────────────────────────

const hostBackend: SandboxBackend = {
  name: "host",
  sandboxed: false,
  run: ({ command, cwd, sandboxConfig, hudAdapter, env }) =>
    runOnHostWithCallbacks(command, cwd, {
      timeoutMs: sandboxConfig.timeoutMs,
      ...(hudAdapter?.callbacks.onStdout ? { onStdout: hudAdapter.callbacks.onStdout } : {}),
      ...(hudAdapter?.callbacks.onStderr ? { onStderr: hudAdapter.callbacks.onStderr } : {}),
      ...(env ? { env } : {}),
    }),
};

const createContainerBackend = (): SandboxBackend => ({
  name: "container",
  sandboxed: true,
  run: ({ command, cwd, sandboxConfig, hudAdapter, env }) =>
    Effect.gen(function* () {
      const credentialMount: CredentialMount | null = yield* createCredentialMount()
        .pipe(
          Effect.provide(BunContext.layer),
          Effect.catchAll((err) => {
            console.warn(`[sandbox] Credential injection skipped: ${err.message}`);
            return Effect.succeed(null);
          }),
        );

      const volumeMounts = credentialMount ? [credentialMount.volumeMount] : [];
      const containerConfig = buildContainerConfig(sandboxConfig, cwd, {
        ...(env ? { env } : {}),
        ...(volumeMounts.length > 0 ? { volumeMounts } : {}),
      });

      const result = yield* runInSandboxContainer(command, containerConfig, hudAdapter?.callbacks).pipe(
        Effect.ensuring(
          credentialMount
            ? cleanupCredentialMount(credentialMount).pipe(
                Effect.provide(BunContext.layer),
                Effect.catchAll(() => Effect.void),
              )
            : Effect.void,
        ),
      );

      return result;
    }),
});

// ─────────────────────────────────────────────────────────────────────────────
// Main Runner API
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Run a shell command, using sandbox if available and enabled.
 *
 * This function:
 * 1. Checks if sandbox is available
 * 2. If available, injects Claude Code credentials and runs command in container
 * 3. If unavailable, falls back to host execution
 * 4. Emits HUD events for real-time streaming to UI
 *
 * @param command - Command as array of strings (e.g., ["bun", "test"])
 * @param config - Sandbox runner configuration
 * @param env - Optional environment variables
 */
export const runCommand = (
  command: string[],
  config: SandboxRunnerConfig,
  env?: Record<string, string>,
): Effect.Effect<CommandResult, Error, never> =>
  Effect.gen(function* () {
    // Validate working directory exists before attempting to run
    // This prevents cryptic container errors when worktrees are corrupted/removed
    if (!nodeFs.existsSync(config.cwd)) {
      config.emit?.({
        type: "sandbox_fallback",
        reason: `Working directory does not exist: ${config.cwd}`,
      });
      return yield* Effect.fail(
        new Error(`Working directory does not exist: ${config.cwd}. The worktree may have been corrupted or removed during execution.`),
      );
    }

    const sandboxAvailable = yield* checkSandboxAvailable(
      config.sandboxConfig,
      config.emit,
    );

    const backend = sandboxAvailable ? createContainerBackend() : hostBackend;
    const hudAdapter = createSandboxHudAdapter(config.emitHud);

    config.emit?.({
      type: "sandbox_command_start",
      command,
      inContainer: backend.sandboxed,
    });

    hudAdapter.emitStart({
      command,
      sandboxed: backend.sandboxed,
      image: backend.sandboxed
        ? config.sandboxConfig.image ?? DEFAULT_SANDBOX_IMAGE
        : "host",
      workdir: config.cwd,
      context: config.context ?? "verification",
    });

  const runWithBackend = (target: SandboxBackend) =>
    target.run({
      command,
      cwd: config.cwd,
      sandboxConfig: config.sandboxConfig,
      ...(env ? { env } : {}),
      hudAdapter,
    });

    const startTime = Date.now();

    const result = yield* (
      backend.sandboxed
        ? runWithBackend(backend).pipe(
            Effect.catchAll((error) => {
              config.emit?.({ type: "sandbox_fallback", reason: error.message });
              hudAdapter.setSandboxed(false);
              return runWithBackend(hostBackend);
            }),
          )
        : runWithBackend(backend)
    );

    const durationMs = Date.now() - startTime;

    config.emit?.({
      type: "sandbox_command_complete",
      command,
      exitCode: result.exitCode,
      durationMs,
    });

    hudAdapter.emitComplete(result.exitCode, durationMs);

    return result;
  });

/**
 * Run a shell command string, using sandbox if available.
 * Parses the command string into arguments for container execution.
 *
 * @param commandString - Shell command string (e.g., "bun test")
 * @param config - Sandbox runner configuration
 * @param env - Optional environment variables
 */
export const runCommandString = (
  commandString: string,
  config: SandboxRunnerConfig,
  env?: Record<string, string>
): Effect.Effect<CommandResult, Error, never> => {
  // Parse command string into array
  // Simple parsing - splits on whitespace (doesn't handle quoted strings perfectly)
  const command = commandString.trim().split(/\s+/);
  return runCommand(command, config, env);
};

// ─────────────────────────────────────────────────────────────────────────────
// Verification Command Runner
// ─────────────────────────────────────────────────────────────────────────────

export interface VerificationResult {
  passed: boolean;
  outputs: string[];
  sandboxed: boolean;
}

/**
 * Run verification commands (typecheck, tests) with sandbox support.
 * This is the main integration point for the orchestrator.
 *
 * @param commands - Array of verification command strings
 * @param config - Sandbox runner configuration
 * @param emit - Optional event emitter for orchestrator events
 */
export const runVerificationWithSandbox = (
  commands: string[],
  config: SandboxRunnerConfig,
  emit?: (event: { type: string; [key: string]: any }) => void
): Effect.Effect<VerificationResult, Error, never> =>
  Effect.gen(function* () {
    const outputs: string[] = [];
    let allPassed = true;
    let anySandboxed = false;

    for (const cmd of commands) {
      emit?.({ type: "verification_start", command: cmd });

      const result = yield* runCommandString(cmd, config);
      const output = result.stdout + (result.stderr ? `\n${result.stderr}` : "");
      outputs.push(output);

      if (result.sandboxed) {
        anySandboxed = true;
      }

      const passed = result.exitCode === 0;
      emit?.({
        type: "verification_complete",
        command: cmd,
        passed,
        output,
        exitCode: result.exitCode,
        sandboxed: result.sandboxed,
      });

      if (!passed) {
        allPassed = false;
      }
    }

    return {
      passed: allPassed,
      outputs,
      sandboxed: anySandboxed,
    };
  });

// ─────────────────────────────────────────────────────────────────────────────
// Exports for Testing
// ─────────────────────────────────────────────────────────────────────────────

export { runOnHost as _runOnHost, runInSandboxContainer as _runInSandboxContainer };

File: crates/agent/src/sandbox_runner.rs
//! Sandbox Runner
//!
//! Provides sandboxed execution for orchestrator operations.
//! When sandbox is enabled, commands run inside a container with the workspace mounted.
//! Falls back to host execution when sandbox is disabled or unavailable.

use crate::error::{AgentError, AgentResult};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/// Sandbox configuration from project.json
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SandboxConfig {
    /// Whether sandbox is enabled
    #[serde(default)]
    pub enabled: Option<bool>,
    /// Backend to use (docker, podman, none, auto)
    #[serde(default)]
    pub backend: Option<String>,
    /// Container image to use
    #[serde(default)]
    pub image: Option<String>,
    /// Memory limit (e.g., "2g")
    #[serde(default)]
    pub memory_limit: Option<String>,
    /// CPU limit (e.g., "2")
    #[serde(default)]
    pub cpu_limit: Option<String>,
    /// Timeout in milliseconds
    #[serde(default)]
    pub timeout_ms: Option<u64>,
}

/// Sandbox runner configuration
#[derive(Debug, Clone)]
pub struct SandboxRunnerConfig {
    /// Sandbox configuration from project.json
    pub sandbox_config: SandboxConfig,
    /// Working directory (repo root) to mount as /workspace
    pub cwd: String,
    /// Execution context for UI grouping (default: "verification")
    pub context: Option<String>,
}

impl Default for SandboxRunnerConfig {
    fn default() -> Self {
        Self {
            sandbox_config: SandboxConfig::default(),
            cwd: ".".to_string(),
            context: Some("verification".to_string()),
        }
    }
}

/// Events emitted during sandbox operations
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SandboxRunnerEvent {
    /// Sandbox availability check started
    SandboxCheckStart,
    /// Sandbox is available
    SandboxAvailable { backend: String },
    /// Sandbox is not available
    SandboxUnavailable { reason: String },
    /// Command execution started
    SandboxCommandStart {
        command: Vec<String>,
        in_container: bool,
    },
    /// Command execution completed
    SandboxCommandComplete {
        command: Vec<String>,
        exit_code: i32,
        duration_ms: u64,
    },
    /// Falling back to host execution
    SandboxFallback { reason: String },
}

/// Result of running a command
#[derive(Debug, Clone)]
pub struct CommandResult {
    /// Exit code
    pub exit_code: i32,
    /// Standard output
    pub stdout: String,
    /// Standard error
    pub stderr: String,
    /// Whether the command ran in a container or on host
    pub sandboxed: bool,
}

/// Container configuration
#[derive(Debug, Clone)]
pub struct ContainerConfig {
    /// Container image
    pub image: String,
    /// Host directory to mount as workspace
    pub workspace_dir: String,
    /// Working directory inside container
    pub workdir: String,
    /// Memory limit
    pub memory_limit: Option<String>,
    /// CPU limit
    pub cpu_limit: Option<String>,
    /// Timeout in milliseconds
    pub timeout_ms: Option<u64>,
    /// Environment variables
    pub env: Option<HashMap<String, String>>,
    /// Volume mounts
    pub volume_mounts: Option<Vec<String>>,
    /// Auto-remove container after exit
    pub auto_remove: bool,
}

/// Default sandbox image
pub const DEFAULT_SANDBOX_IMAGE: &str = "oven/bun:latest";

// ─────────────────────────────────────────────────────────────────────────────
// Sandbox Availability Check
// ─────────────────────────────────────────────────────────────────────────────

/// Check if a container backend (docker/podman) is available
pub fn is_container_available() -> bool {
    // Try docker first
    if Command::new("docker")
        .args(["info"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
    {
        return true;
    }

    // Try podman
    Command::new("podman")
        .args(["info"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

/// Get the available container backend
pub fn get_container_backend() -> Option<String> {
    if Command::new("docker")
        .args(["info"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
    {
        return Some("docker".to_string());
    }

    if Command::new("podman")
        .args(["info"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
    {
        return Some("podman".to_string());
    }

    None
}

/// Check if sandbox execution is available
pub fn check_sandbox_available(config: &SandboxConfig) -> (bool, Option<SandboxRunnerEvent>) {
    // Check if explicitly disabled
    if config.enabled == Some(false) {
        return (
            false,
            Some(SandboxRunnerEvent::SandboxUnavailable {
                reason: "sandbox.enabled is false".to_string(),
            }),
        );
    }

    // Check if backend is explicitly "none"
    if config.backend.as_deref() == Some("none") {
        return (
            false,
            Some(SandboxRunnerEvent::SandboxUnavailable {
                reason: "sandbox.backend is 'none'".to_string(),
            }),
        );
    }

    // Check if container backend is available
    if is_container_available() {
        let backend = config
            .backend
            .clone()
            .unwrap_or_else(|| "auto".to_string());
        (
            true,
            Some(SandboxRunnerEvent::SandboxAvailable { backend }),
        )
    } else {
        (
            false,
            Some(SandboxRunnerEvent::SandboxUnavailable {
                reason: "no container backend available".to_string(),
            }),
        )
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Container Config Builder
// ─────────────────────────────────────────────────────────────────────────────

/// Build ContainerConfig from SandboxConfig and working directory
pub fn build_container_config(
    sandbox_config: &SandboxConfig,
    cwd: &str,
    env: Option<HashMap<String, String>>,
    volume_mounts: Option<Vec<String>>,
) -> ContainerConfig {
    ContainerConfig {
        image: sandbox_config
            .image
            .clone()
            .unwrap_or_else(|| DEFAULT_SANDBOX_IMAGE.to_string()),
        workspace_dir: cwd.to_string(),
        workdir: "/workspace".to_string(),
        memory_limit: sandbox_config.memory_limit.clone(),
        cpu_limit: sandbox_config.cpu_limit.clone(),
        timeout_ms: sandbox_config.timeout_ms,
        env,
        volume_mounts,
        auto_remove: true,
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Host Execution
// ─────────────────────────────────────────────────────────────────────────────

/// Run a command on the host (no sandbox)
pub fn run_on_host(
    command: &[String],
    cwd: &str,
    timeout_ms: Option<u64>,
    env: Option<&HashMap<String, String>>,
) -> AgentResult<CommandResult> {
    // Validate working directory exists
    if !Path::new(cwd).exists() {
        return Err(AgentError::Io(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!("Working directory does not exist: {}", cwd),
        )));
    }

    let cmd_string = command.join(" ");
    let timeout = Duration::from_millis(timeout_ms.unwrap_or(120_000));
    let start = Instant::now();

    let mut cmd = Command::new("sh");
    cmd.args(["-c", &cmd_string])
        .current_dir(cwd)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    // Add environment variables
    if let Some(env_vars) = env {
        for (key, value) in env_vars {
            cmd.env(key, value);
        }
    }

    let output = cmd.output().map_err(AgentError::Io)?;

    let elapsed = start.elapsed();
    if elapsed > timeout {
        return Err(AgentError::Timeout(format!(
            "Host command timed out after {}ms",
            timeout.as_millis()
        )));
    }

    Ok(CommandResult {
        exit_code: output.status.code().unwrap_or(1),
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
        sandboxed: false,
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// Sandboxed Execution
// ─────────────────────────────────────────────────────────────────────────────

/// Run a command in a sandbox container
pub fn run_in_container(
    command: &[String],
    config: &ContainerConfig,
) -> AgentResult<CommandResult> {
    let backend = get_container_backend().ok_or_else(|| {
        AgentError::ContainerError("No container backend (docker/podman) available".to_string())
    })?;

    let mut args = vec!["run".to_string(), "--rm".to_string()];

    // Add workdir
    args.push("-w".to_string());
    args.push(config.workdir.clone());

    // Mount workspace
    args.push("-v".to_string());
    args.push(format!("{}:{}", config.workspace_dir, config.workdir));

    // Add memory limit
    if let Some(ref limit) = config.memory_limit {
        args.push("-m".to_string());
        args.push(limit.clone());
    }

    // Add CPU limit
    if let Some(ref limit) = config.cpu_limit {
        args.push("--cpus".to_string());
        args.push(limit.clone());
    }

    // Add environment variables
    if let Some(ref env) = config.env {
        for (key, value) in env {
            args.push("-e".to_string());
            args.push(format!("{}={}", key, value));
        }
    }

    // Add volume mounts
    if let Some(ref mounts) = config.volume_mounts {
        for mount in mounts {
            args.push("-v".to_string());
            args.push(mount.clone());
        }
    }

    // Add image
    args.push(config.image.clone());

    // Add command
    args.push("sh".to_string());
    args.push("-c".to_string());
    args.push(command.join(" "));

    let timeout = Duration::from_millis(config.timeout_ms.unwrap_or(120_000));
    let start = Instant::now();

    let output = Command::new(&backend)
        .args(&args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .map_err(AgentError::Io)?;

    let elapsed = start.elapsed();
    if elapsed > timeout {
        return Err(AgentError::Timeout(format!(
            "Container command timed out after {}ms",
            timeout.as_millis()
        )));
    }

    Ok(CommandResult {
        exit_code: output.status.code().unwrap_or(1),
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
        sandboxed: true,
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// Main Runner API
// ─────────────────────────────────────────────────────────────────────────────

/// Run a shell command, using sandbox if available and enabled
pub fn run_command(
    command: &[String],
    config: &SandboxRunnerConfig,
    env: Option<&HashMap<String, String>>,
) -> AgentResult<CommandResult> {
    // Validate working directory exists
    if !Path::new(&config.cwd).exists() {
        return Err(AgentError::Io(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!(
                "Working directory does not exist: {}. The worktree may have been corrupted or removed during execution.",
                config.cwd
            ),
        )));
    }

    let (sandbox_available, _event) = check_sandbox_available(&config.sandbox_config);

    if sandbox_available {
        // Try container execution with fallback to host
        let container_config = build_container_config(
            &config.sandbox_config,
            &config.cwd,
            env.cloned(),
            None,
        );

        match run_in_container(command, &container_config) {
            Ok(result) => Ok(result),
            Err(_) => {
                // Fallback to host
                run_on_host(
                    command,
                    &config.cwd,
                    config.sandbox_config.timeout_ms,
                    env,
                )
            }
        }
    } else {
        run_on_host(
            command,
            &config.cwd,
            config.sandbox_config.timeout_ms,
            env,
        )
    }
}

/// Run a shell command string, using sandbox if available
pub fn run_command_string(
    command_string: &str,
    config: &SandboxRunnerConfig,
    env: Option<&HashMap<String, String>>,
) -> AgentResult<CommandResult> {
    let command: Vec<String> = command_string
        .trim()
        .split_whitespace()
        .map(|s| s.to_string())
        .collect();
    run_command(&command, config, env)
}

// ─────────────────────────────────────────────────────────────────────────────
// Verification Command Runner
// ─────────────────────────────────────────────────────────────────────────────

/// Result of running verification commands
#[derive(Debug, Clone)]
pub struct SandboxVerificationResult {
    /// Whether all verification commands passed
    pub passed: bool,
    /// Output from each verification command
    pub outputs: Vec<String>,
    /// Whether any command ran in a container
    pub sandboxed: bool,
}

/// Run verification commands (typecheck, tests) with sandbox support
pub fn run_verification_with_sandbox(
    commands: &[String],
    config: &SandboxRunnerConfig,
) -> AgentResult<SandboxVerificationResult> {
    let mut outputs = Vec::new();
    let mut all_passed = true;
    let mut any_sandboxed = false;

    for cmd in commands {
        let result = run_command_string(cmd, config, None)?;
        let output = if result.stderr.is_empty() {
            result.stdout.clone()
        } else {
            format!("{}\n{}", result.stdout, result.stderr)
        };
        outputs.push(output);

        if result.sandboxed {
            any_sandboxed = true;
        }

        if result.exit_code != 0 {
            all_passed = false;
        }
    }

    Ok(SandboxVerificationResult {
        passed: all_passed,
        outputs,
        sandboxed: any_sandboxed,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sandbox_config_default() {
        let config = SandboxConfig::default();
        assert!(config.enabled.is_none());
        assert!(config.backend.is_none());
        assert!(config.image.is_none());
    }

    #[test]
    fn test_check_sandbox_available_disabled() {
        let config = SandboxConfig {
            enabled: Some(false),
            ..Default::default()
        };
        let (available, event) = check_sandbox_available(&config);
        assert!(!available);
        assert!(matches!(
            event,
            Some(SandboxRunnerEvent::SandboxUnavailable { .. })
        ));
    }

    #[test]
    fn test_check_sandbox_available_backend_none() {
        let config = SandboxConfig {
            backend: Some("none".to_string()),
            ..Default::default()
        };
        let (available, event) = check_sandbox_available(&config);
        assert!(!available);
        assert!(matches!(
            event,
            Some(SandboxRunnerEvent::SandboxUnavailable { reason })
            if reason.contains("none")
        ));
    }

    #[test]
    fn test_build_container_config() {
        let sandbox = SandboxConfig {
            image: Some("test:latest".to_string()),
            memory_limit: Some("1g".to_string()),
            ..Default::default()
        };
        let config = build_container_config(&sandbox, "/workspace", None, None);

        assert_eq!(config.image, "test:latest");
        assert_eq!(config.workspace_dir, "/workspace");
        assert_eq!(config.workdir, "/workspace");
        assert_eq!(config.memory_limit, Some("1g".to_string()));
        assert!(config.auto_remove);
    }

    #[test]
    fn test_build_container_config_defaults() {
        let sandbox = SandboxConfig::default();
        let config = build_container_config(&sandbox, "/work", None, None);

        assert_eq!(config.image, DEFAULT_SANDBOX_IMAGE);
    }

    #[test]
    fn test_command_result_fields() {
        let result = CommandResult {
            exit_code: 0,
            stdout: "output".to_string(),
            stderr: "".to_string(),
            sandboxed: true,
        };

        assert_eq!(result.exit_code, 0);
        assert_eq!(result.stdout, "output");
        assert!(result.sandboxed);
    }

    #[test]
    fn test_sandbox_runner_event_serialization() {
        let event = SandboxRunnerEvent::SandboxAvailable {
            backend: "docker".to_string(),
        };
        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains("sandbox_available"));
        assert!(json.contains("docker"));
    }

    #[test]
    fn test_run_on_host_missing_dir() {
        let result = run_on_host(
            &["echo".to_string(), "hello".to_string()],
            "/nonexistent/path",
            None,
            None,
        );
        assert!(result.is_err());
    }

    #[test]
    fn test_run_on_host_simple() {
        let result = run_on_host(
            &["echo".to_string(), "hello".to_string()],
            ".",
            None,
            None,
        );
        assert!(result.is_ok());
        let res = result.unwrap();
        assert_eq!(res.exit_code, 0);
        assert!(res.stdout.contains("hello"));
        assert!(!res.sandboxed);
    }

    #[test]
    fn test_run_command_string() {
        let config = SandboxRunnerConfig {
            sandbox_config: SandboxConfig {
                enabled: Some(false),
                ..Default::default()
            },
            cwd: ".".to_string(),
            context: None,
        };

        let result = run_command_string("echo test", &config, None);
        assert!(result.is_ok());
        let res = result.unwrap();
        assert_eq!(res.exit_code, 0);
        assert!(res.stdout.contains("test"));
    }

    #[test]
    fn test_sandbox_verification_result() {
        let result = SandboxVerificationResult {
            passed: true,
            outputs: vec!["ok".to_string()],
            sandboxed: false,
        };
        assert!(result.passed);
        assert_eq!(result.outputs.len(), 1);
    }

    #[test]
    fn test_run_verification_with_sandbox() {
        let config = SandboxRunnerConfig {
            sandbox_config: SandboxConfig {
                enabled: Some(false),
                ..Default::default()
            },
            cwd: ".".to_string(),
            context: None,
        };

        let commands = vec!["echo ok".to_string()];
        let result = run_verification_with_sandbox(&commands, &config);
        assert!(result.is_ok());
        let res = result.unwrap();
        assert!(res.passed);
        assert!(!res.sandboxed);
    }
}
